src/rtp/ext.rs

```diff
@@ -635,8 +710,95 @@ pub struct ExtensionValues {
     pub mid: Option<Mid>,
     #[doc(hidden)]
     pub frame_mark: Option<u32>,
+
+    /// User values for [`ExtensionSerializer`] to parse into and write from.
+    pub user_values: UserExtensionValues,
+}
+
+// The "AnyMap" idea is borrowed from the http crate but replacing Box for Any.
+type AnyMap = HashMap<TypeId, Arc<dyn Any + Send + Sync>, BuildHasherDefault<IdHasher>>;
+
+// No point in hashing the TypeId, since it is already unique.
+#[derive(Default)]
+struct IdHasher(u64);
+
+impl Hasher for IdHasher {
+    fn write(&mut self, _: &[u8]) {
+        unreachable!("TypeId calls write_u64");
+    }
+
+    #[inline]
+    fn write_u64(&mut self, id: u64) {
+        self.0 = id;
+    }
+
+    #[inline]
+    fn finish(&self) -> u64 {
+        self.0
+    }
 }
 
+/// Space for storing user extension values via [`ExtensionSerializer`].
+#[derive(Clone, Default)]
+pub struct UserExtensionValues {
+    map: Option<AnyMap>,
+}
+
+// TODO: I don't see a good way of comparing this. Is there one?
+impl PartialEq for UserExtensionValues {
+    fn eq(&self, _other: &Self) -> bool {
+        true
+    }
+}
+
+impl Eq for UserExtensionValues {}
+
+impl UserExtensionValues {
+    /// Set a user extension value.
+    ///
+    /// This uses the type of the value as "key", i.e. it can only hold a single
+    /// per type. The user should make a wrapper type for the extension they want
+    /// to parse/write.
+    ///
+    /// ```
+    /// # use str0m::rtp::ExtensionValues;
+    /// let mut exts = ExtensionValues::default();
+    ///
+    /// #[derive(Debug, PartialEq, Eq)]
+    /// struct MySpecialType(u8);
+    ///
+    /// exts.user_values.set(MySpecialType(42));
+    /// ```
+    pub fn set<T: Send + Sync + 'static>(&mut self, val: T) {
+        self.map
+            .get_or_insert_with(HashMap::default)
+            .insert(TypeId::of::<T>(), Arc::new(val));
+    }
+
+    /// Get a user extension value (by type).
+    /// ```
+    /// # use str0m::rtp::ExtensionValues;
+    /// let mut exts = ExtensionValues::default();
+    ///
+    /// #[derive(Debug, PartialEq, Eq)]
+    /// struct MySpecialType(u8);
+    ///
+    /// exts.user_values.set(MySpecialType(42));
+    ///
+    /// let v = exts.user_values.get::<MySpecialType>();
+    ///
+    /// assert_eq!(v, Some(&MySpecialType(42)));
+    /// ```
+    pub fn get<T: Send + Sync + 'static>(&self) -> Option<&T> {
+        self.map
+            .as_ref()
+            .and_then(|map| map.get(&TypeId::of::<T>()))
+            .and_then(|boxed| (&**boxed as &(dyn Any + 'static)).downcast_ref())
```

Comment:
Isn't this downcast succeeding an invariant?

Reply:
Might be. That code is lifted straight from http crate, I don't know if they thought about something I didn't.

Reply:
A difference stance on panics. The only way to get a value into the map is via `set` and there the invariant that the value stored for the key `TypeId::of::<T>()` is a `T` is enforced. Therefore, the value for a given `TypeId::of::<T>()` should always be a `T`. I don't think `downcast_ref` will ever fail if this invariant is upheld.

https://doc.rust-lang.org/1.72.0/src/core/any.rs.html#289-293

Reply:
Alright. I'll do a force unwrap()